cmd/kubeadm/app/phases/upgrade/postupgrade.go

```diff
@@ -119,8 +120,15 @@ func removeOldDNSDeploymentIfAnotherDNSIsUsed(cfg *kubeadmapi.ClusterConfigurati
 			deploymentToDelete = kubeadmconstants.KubeDNSDeploymentName
 		}
 
-		// If we're dry-running, we don't need to wait for the new DNS addon to become ready
-		if !dryRun {
+		nodes, err := client.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{
+			FieldSelector: fields.Set{"spec.unschedulable": "false"}.AsSelector().String(),
```

Comment:
i do not recall observing nodes being unschedulable while the CoreDNS addon is being upgraded during "kubeadm upgrade". is this something you have seen @SataQiu ?

Reply:
According to this guide https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/#upgrade-the-first-control-plane-node, `kubectl drain <cp-node-name> --ignore-daemonsets` in step 2 will make the control plane node unschedulable.
If we are using a single node cluster, the only node will be marked as unschedulable. In this case, new DNS deployment will never be ready. That's why the program is stuck.

Reply:
that is true.